plugins/uefi/fu-plugin-uefi.c

```diff
@@ -613,9 +612,63 @@ load_custom_esp (FuPlugin *plugin, GError **error)
 
 			return FALSE;
 		}
+#ifdef HAVE_FWUP_CUSTOM_ESP
+		fwup_set_esp_mountpoint (data->esp_path);
 		g_debug ("%s set to %s", key, data->esp_path);
-		fu_plugin_add_report_metadata (plugin, key,
-					       data->esp_path);
+#endif
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_uefi_delete_old_capsules (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	const gchar *pattern = "/boot/efi/EFI/*/fw/fwupdate-*.cap";
+	g_autoptr(GPtrArray) files = NULL;
+
+	/* delete any files matching the glob in the ESP */
+	files = fu_common_get_files_recursive (data->esp_path, error);
+	if (files == NULL)
+		return FALSE;
+	for (guint i = 0; i < files->len; i++) {
+		const gchar *fn = g_ptr_array_index (files, i);
+		if (fnmatch (pattern, fn, 0) == 0) {
+			g_autoptr(GFile) file = g_file_new_for_path (fn);
+			g_debug ("deleting %s", fn);
+			if (!g_file_delete (file, NULL, error))
+				return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+gboolean
+fu_plugin_startup (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	/* load any overriden options */
+	if (!fu_plugin_uefi_set_custom_mountpoint (plugin, error))
+		return FALSE;
+
+	/* get the default compiled-in value for the ESP mountpoint */
+#ifdef HAVE_FWUP_GET_ESP_MOUNTPOINT
+	if (data->esp_path == NULL)
+		data->esp_path = g_strdup (fwup_get_esp_mountpoint ());
+#endif
+
+	/* delete any existing .cap files to avoid the small ESP partition
+	 * from running out of space when we've done lots of firmware updates */
+	if (data->esp_path != NULL) {
```

Comment:
You'll need to differentiate between boots too. Otherwise restarting fwupd will cause you to possibly lose staged updates.

Reply:
I didn't think about this. We don't have any persistent state we can use for comparing the current boot time to the time this was scheduled. We could use a heuristic e.g. `time_now - boot_time < 60s` or something?

Reply:
Originally you were talking about only running this code when deemed success. Another way to check might be to check UEFI results and see if they're marked succesful from last attempt and make sure last_attempt_version matched what current version is on the system.

Reply:
I did think about restricting this to success, but we don't know what the .cap random file name is going to be on the ESP for a given history item.

Reply:
This should be determinable by parsing efivars fwupdate-$(ESRT)-$(GLOBAL_GUID).

Reply:
> Originally you were talking about only running this code when deemed success.

Yes, but then I wondered what benefit there was in keeping them around on failure. I came to the conclusion that the user was highly unlikely to reschedule the update by changing the BootNext in efibootmgr, and was far more likely just to type `fwupdmgr update` again. In the latter case we already have the *.cab* in the download cache, and so deploying the *.cap* to the ESP again is trivial. It also means we don't have to worry if fwupdate chooses a different unique name for the same firmware, or if something has already deleted the fwupdate-* efivars before fwupd started.

Reply:
So poor mans way to tell if you've rebooted - check for the BootNext variable and if it points to an entry that resembles some "Linux Firmware Updater".  If there is a Bootnext entry and it does you're probably waiting around for a reboot.

I suppose since fwupdate does reuse the variable you can keep it around.  I am just thinking about that scenario of reuse with changing distros.  It's not really that common, and maybe the right hammer there really is to make sure the cleanup script is getting run.